components/launcher-client/src/client.rs

```diff
@@ -75,6 +95,46 @@ impl LauncherCli {
         }
     }
 
+    /// EXPERIMENTAL! BEWARE! EXPERIMENTAL!
+    ///
+    /// When this was added, we needed a way for a Supervisor to be
+    /// able to recover from sending a message to an older version of
+    /// the Launcher that didn't know how to process that message. At
+    /// the time, the Launcher would just ignore the message, but we
+    /// always use a `recv` to wait for the response, so the
+    /// Supervisor would hang.
+    ///
+    /// This function is an attempt to get around this situation, but
+    /// ONLY for the specific scenarios where it is appropriate. I am
+    /// hesitant to introduce a global timeout at this point, because
+    /// it's difficult to know how that will affect the overall
+    /// interactions between the Supervisor and the Launcher (it
+    /// *should* be fine, but I can't guarantee that right now).
+    ///
+    /// As such, use this with caution and intention.
+    fn recv_timeout<T>(rx: &IpcReceiver<Vec<u8>>, timeout: Duration) -> Result<T>
```

Comment:
Just thinking out loud... If we had a way to check the version of the launcher, would this be needed? Unfortunately, if this was accomplished through an API call to the launcher, it would not help in this specific scenario (because the old version of the launcher will not have the API), but maybe it would be worth adding for future situations like this?

Reply:
That would certainly have helped, but unfortunately we don't have such an API right now :frowning:

This particular situation was complicated by the fact that if you sent a message to a Launcher that didn't know how to process that message, the Launcher would just ignore it. However, the Supervisor was always using a blocking `recv` call, so it would hang waiting for a reply that was never going to come. The end result is that you can basically get yourself into a mutual deadlock situation between both processes.

Given all that, this seemed the most straightforward way forward (this PR _does_ fix that "ignore unknown messages" issue, though).

Interestingly, the Launcher _does_ have the ability to see what version of the Supervisor it's running. Adding that ability to the Supervisor seems like a good idea.